src/test_regex.cpp

```diff
@@ -0,0 +1,154 @@
+            "prompt    : used copilot",
+            "prompt:       USed copilot",
+            "prompt:used copilot",
+            "prompt: used copilot     ",
+            "     prompt: used copilot     ",
+            "// prompt: used copilot",
+            "// prompt: used     coPilot",
+            "// prompt    : used Copilot",
+            "//prompt:used copilot",
+            "// prompt:used copilot",
+            "//prompt:    used copilot",
+            "// prompt:    used copilot",
+            "     // prompt: used copilot",
+            "     // prompt: used copilot      ",
+            "     // PromPt: used copilot      ",
+            "prompt: https://www.google.com/",
+            "prompt: https://www.google.com",
+            "prompt    : https://www.google.com/",
+            "prompt:       https://www.google.com/",
+            "prompt:https://www.google.com/",
+            "prompt: https://www.google.com/     ",
+            "     prompt: https://www.google.com/     ",
+            "// prompt: https://www.google.com/",
+            "// prompt: https://www.google.com",
+            "// prompt    : https://www.google.com/",
+            "//prompt:https://www.google.com/",
+            "// prompt:https://www.google.com/",
+            "//prompt:    https://www.google.com/",
+            "// prompt:    https://www.google.com/",
+            "     // prompt: https://www.google.com/",
+            "     // prompt: https://www.google.com/      ",
+            "     // PromPt: https://www.google.com/      "
+    };
+
+    for (const auto& test_str : test_strings) {
+        std::smatch match;
+        std::cout << "Test string: " << test_str << std::endl;
+        if (std::regex_search(test_str, match, pattern) && match.size() > 1) {
+            std::string prompt = match[1].str();
+            // Remove all the whitespace characters from both ends of the string
+            prompt = std::regex_replace(prompt, std::regex("^\\s+|\\s+$"), "");
+            std::regex urlRegex(R"(^https?:\/\/[^\s\/$.?#].[^\s]*$)", std::regex_constants::icase);
+            if (!std::regex_match(prompt, urlRegex)) {
+                prompt = toLowercase(prompt);
+                // Replace more than one whitespace characters with one whitespace character only
+                prompt = std::regex_replace(prompt, std::regex(R"(\s+)"), " ");
+            }
+            std::cout << "Prompt: " << prompt << std::endl;
+            std::cout << "Match found: " << test_str << std::endl;
+        } else {
+            std::cout << "No match: " << test_str << std::endl;
+        }
+        std::cout << std::string(20, '=') << std::endl;
+    }
+}
+
+void testEndTagRegex() {
+    std::regex pattern(R"(ai-gen\s*end)", std::regex_constants::icase);
+
+    std::string test_strings[] = {
+            "ai-gen end",
+            "ai-gen    end",
+            "ai-gen end     ",
+            "     ai-gen end",
+            "     ai-gen end     ",
```

Comment:
This is invalid syntax.
The text needs to be a comment, otherwise the compiler will throw an error, maybe can remove these test cases?

Reply:
Yes, you're right, this might not be very useful. Additionally, I'm actually considering the generalizability of regular expressions, because I'm thinking about the possibility of students making typos or any unforeseen mistakes. If it could match as many cases as possible, that would be even better. Therefore, I'm also wondering if there might be an algorithm that can match some cases of typos, for example:
- // ai-gen end
- // ai-gen      end
- //ai-gen end
- // ai gen end
- // ai -gen end
- // ai gen-end
- // ai-gen End
- // AI-GEN END
- // ai-gan end
etc.

However, this might not be very important and can be considered later, haha.